### ISSUE TITLE
Client certificate with passwordless login

### ISSUE THREAD

**Issue by christopherL91** [erlio/vernemq#758]
### Environment

Hi! Thank you for your excellent work and for bringing a nice piece of software to this world.
I'm having a small issue regarding client certificate authorization that you guys might have a solution for.
I'm running haproxy with proxy protocol turned on and I'm verifying the certificate inside haproxy, the username is automatically being set to the CN as specified inside the certificate and that works from the vernemq side. However, I have an ACL file that I don't think is being respected. I would like to have passwordless login (the client brings a unique client certificate so I think passwords are being redundant)

The ACL file looks like this:

```
user backend
topic write all

user node
topic read node/gw/+/cmd
topic read node/wb/+/cmd
```

The client certificate has CN set to "node" and I can confirm I'm being able to read those topics, however, `allow_anonymous` is set to true (to enable passwordless logins) but then a user is able to read/write to all topics, side stepping the ACL file (I have confirmed this with mosquitto_sub/mosquitto_pub) How can I solve this issue?

tldr;

enable passwordless login with client certificate while setting a ACL file that is respected.

- VerneMQ Version:
1.4.0

- OS:
debian (container)
- Erlang/OTP version (if building from source):

### Expected behavior

- Please describe the expected behaviour.

When connecting with a client certificate the username is set to the value of the CN, this works. However the ACL file is being side stepped when `allow_anonymous` is set to true. My goal is to enabled passwordless login (the client brings a certificate, that is being terminated in haproxy)
I would expect the software to block any other publish/subscribe calls that isn't explicitly set in the ACL file.

### Actual behaviour

- Please include minimal steps to reproduce the issue, if possible.

The ACL file seems to be ignored when `allow_anonymous` is set. I can confirm that the session belongs to the user by running `vmq-admin sessions show`

**Comment by christopherL91** [comment]
@ioolkos thanks for being so quick!

Yes I have read it, but I still can't figure out why the ACL parts is affected. The relevant parts are:

```
allow_anonymous = on
plugins.vmq_passwd = off
plugins.vmq_acl = on
```

```
user backend
topic write all

user node
topic read node/gw/+/cmd
topic read node/wb/+/cmd
```

```
mosquitto_sub -h REDACTED -p 8883 -t node/gw/d/cmd --cafile cacerts.pem --cert client.pem --key client-key.pem -d
```

When issuing the above call, I can see that a session is created with the username `node` just as intended. However, I'm allowed to change the topic part to whatever I want, and vernemq allows that for some reason. What am I missing?

**Comment by ioolkos** [comment]
Hm, thanks. I don't see you doing anything wrong here in any case. 
We'll have to check & understand this.

**Comment by ioolkos** [comment]
@dergraf, @larshesel Username can be of form username() or {preauth, string()}. Second is for the case where the username comes in already authenticated as in client certs.
https://github.com/erlio/vernemq/blob/a610cc770e67d4f439dd4488d64714b1d2b73231/apps/vmq_server/src/vmq_mqtt_fsm.erl#L44

Reminder to check what the FSM does here with the 'preauth' case, especially whether the authorization chain is correct.

**Comment by dergraf** [comment]
@ioolkos just checked.. the case with a `{preauth, Username}` and `allow_anonymous=on` should work, actually the extraction of the username from the preauth tuple happens before checking going through the authentication steps (which includes checking for `allow_anonymous`).
See: https://github.com/erlio/vernemq/blob/a610cc770e67d4f439dd4488d64714b1d2b73231/apps/vmq_server/src/vmq_mqtt_fsm.erl#L487

**Comment by dergraf** [comment]
@christopherL91 what is the status of this issue, have you been able to solve it?
